lib/compat/wordpress-6.3/script-loader.php

```diff
@@ -83,3 +83,107 @@ function( $settings ) {
+	// Remove the block styles from the footer.
+	remove_action( 'wp_footer', 'wp_maybe_inline_styles', 1 );
+}
+
+/**
+ * Print block styles inline for non-block (classic) themes.
+ * Improves performance for classic themes by only loading the styles that are needed
+ * for the blocks that are actually used on the page, instead of loading all block styles.
+ * At the same time, this reduces layout shifts by inlining these styles
+ * instead of printing them all in the footer.
+ *
+ * NOTE: When backported to Core,
+ * this should probably be merged with the `wp_maybe_inline_styles` function.
+ *
+ * This function is hooked on `render_block`.
+ *
+ * @param string $block_content The block content.
+ * @param array  $block         The full block, including name and attributes.
+ *
+ * @return string The block content.
+ */
+function gutenberg_inline_block_styles_on_classic_themes( $block_content, $block ) {
+	// Bail early if not a classic theme.
+	if ( wp_is_block_theme() ) {
+		return $block_content;
+	}
+
+	// Get all registered blocks. Use a static var to avoid calling this
+	// for each block that gets rendered on the page.
+	static $all_registered_blocks;
+	if ( ! $all_registered_blocks ) {
+		$all_registered_blocks = WP_Block_Type_Registry::get_instance()->get_all_registered();
+	}
+
+	$style_handles = array();
+	if ( ! empty( $all_registered_blocks[ $block['blockName'] ] ) ) {
+		$style_handles = $all_registered_blocks[ $block['blockName'] ]->style_handles;
+	}
+
+	// Bail early if there are no styles for this block.
+	if ( empty( $style_handles ) ) {
+		return $block_content;
+	}
+
+	/* This filter is documented in wp-includes/script-loader.php */
+	$total_inline_limit = apply_filters( 'styles_inline_size_limit', 20000 );
+
+	// A static var to keep track of the total inlined CSS size.
+	static $total_inline_size = 0;
+
+	// Get the WP_Styles object.
+	$wp_styles = wp_styles();
+
+	foreach ( $style_handles as $style_handle ) {
+		// Skip stylesheet if it has already been printed.
+		if ( in_array( $style_handle, $wp_styles->done, true ) ) {
+			continue;
+		}
+
+		$path  = wp_styles()->get_data( $style_handle, 'path' );
+		$after = wp_styles()->get_data( $style_handle, 'after' );
+		if ( is_array( $after ) ) {
+			$after = implode( '', $after );
+		}
+
+		// Skip stylesheet if it has no path or "after" data.
+		if ( ! $path && ! $after ) {
+			continue;
+		}
+
+		// Get the CSS for this stylesheet.
+		$css = file_exists( $path ) ? file_get_contents( $path ) : '';
+
+		// Add the "after" styles.
+		$css .= $after;
+
+		// Get the size of the CSS.
+		$styles_size = strlen( $css );
+
+		// Skip if styles are empty.
+		if ( 0 === $styles_size ) {
+			$wp_styles->done[] = $style_handle;
+			continue;
+		}
+
+		// Check if adding these styles inline will exceed the limit.
+		// If it does, then bail early.
+		if ( $total_inline_size + $styles_size > $total_inline_limit ) {
+			continue;
+		}
```

Comment:
Doing this here is problematic. How we need to handle this with classic themes is a bit different from how it works with block themes: If we just skip any styles here that exceed the threshold, they end up being loaded only in the footer, which increases the risk of layout shifts. To avoid that while still respecting the threshold, we would need to output the external stylesheet (`<link rel="stylesheet" ...`) in here still, i.e. before each block's content the first time that it is used.

While that is perfectly valid HTML5, maybe that takes it a bit far, at least for a V1. I see the following options for us here:
* Either we just go for it, i.e. print `link` tags to load the external stylesheets before the block content as well, i.e. we end up with having `link` tags within the content in case where the amount of CSS exceeds the threshold.
* Alternatively, we don't use the threshold at all for classic themes, effectively ignoring it. Since this entire behavior from this function here can be opted in or opted out (via `should_load_separate_core_block_assets`), I'm not sure the threshold is really valuable in the first place here. If a site maintainer thinks the inline CSS is too much, they don't have to use this behavior entirely.
* Yet another solution would be to simply accept the increased risk for layout shifts in that scenario. In any case, it's safe to say this PR already makes it better than how it is today, so overall this PR still means a reduced risk for CLS for any sites with classic themes that have `should_load_separate_core_block_assets` set to `true`.